src/System.Private.CoreLib/shared/System/Number.DiyFp.cs

```diff
@@ -8,121 +8,141 @@ namespace System
+            }
 
-                set
-                {
-                    _e = value;
-                }
+            public DiyFp(ulong significand, int exponent)
+            {
+                f = significand;
+                e = exponent;
             }
 
-            public static void GenerateNormalizedDiyFp(double value, out DiyFp result)
+            public DiyFp Multiply(in DiyFp other)
             {
-                Debug.Assert(value > 0.0);
+                // Simply "emulates" a 128-bit multiplcation
+                //
+                // However: the resulting number only contains 64-bits. The least
+                // signficant 64-bits are only used for rounding the most significant
+                // 64-bits.
 
-                long f = ExtractFractionAndBiasedExponent(value, out int e);
+                uint a = (uint)(f >> 32);
+                uint b = (uint)(f);
 
-                while ((f & (1L << 52)) == 0)
-                {
-                    f <<= 1;
-                    e--;
-                }
+                uint c = (uint)(other.f >> 32);
+                uint d = (uint)(other.f);
+
+                ulong ac = ((ulong)(a) * c);
+                ulong bc = ((ulong)(b) * c);
+                ulong ad = ((ulong)(a) * d);
+                ulong bd = ((ulong)(b) * d);
 
-                int lengthDelta = (SignificandLength - 53);
-                f <<= lengthDelta;
-                e -= lengthDelta;
+                ulong tmp = (bd >> 32) + (uint)(ad) + (uint)(bc);
 
-                result = new DiyFp((ulong)(f), e);
-            }
+                // By adding (1UL << 31) to tmp, we round the final result.
+                // Halfway cases will be rounded up.
 
-            public static void Minus(ref DiyFp lhs, ref DiyFp rhs, out DiyFp result)
-            {
-                result = lhs;
-                result.Minus(ref rhs);
+                tmp += (1U << 31);
+
+                return new DiyFp((ac + (ad >> 32) + (bc >> 32) + (tmp >> 32)), (e + other.e + SignificandSize));
             }
 
-            public static void Multiply(ref DiyFp lhs, ref DiyFp rhs, out DiyFp result)
+            public DiyFp Normalize()
             {
-                result = lhs;
-                result.Multiply(ref rhs);
+                // This method is mainly called for normalizing boundaries.
+                //
+                // We deviate from the reference implementation by just using
+                // our LeadingZeroCount function so that we only need to shift
+                // and subtract once.
+
+                Debug.Assert(f != 0);
+                int lzcnt = (int)(BigInteger.LeadingZeroCount(f));
+                return new DiyFp((f << lzcnt), (e - lzcnt));
             }
 
-            public void Minus(ref DiyFp rhs)
+            // The exponents of both numbers must be the same.
+            // The significand of this must be bigger than the significand of other.
```

Comment:
`... of this must...` awkward to read, seems like a word is missing, but you are referring to `this` instance and `other`.